posts/convenient-secrets-with-spring-boot-cloud-aws.md

```diff
@@ -0,0 +1,106 @@
+---
+authors:
+- John Strickler
+tags:
+- Secrets Management
+- Spring Boot
+- Spring Cloud
+- AWS Parameter Store
+- Elastic Container Service
+date: 2020-07-07T00:00:00.000Z
+title: "Secret Management w/ Spring Cloud and AWS Parameter Store"
+image: 2020/07/screen-web-design-developing-codes-1936299.jpg
+---
+
+Today's applications are broken into smaller and smaller pieces.  We've been slowly transitioning away from managing our own infrastructure, from using virtual instances, to deploying to clusters.  Your environment may have one or more clusters.  Those clusters have many services.  Those services have many tasks.  Each task is an application running in its own container.  And more than likely, that application requires **configuration**.   
+
+I want to share with you a simple way to store your sensitive configuration and to inject that configuration into your application.  This solution is appealing because there's no infrastructure to manage, is very low maintenance, and scales beautifully.  So let's get started.
+
+# Storing your configuration
+
+For a contrived example, we'll store a *GitHub Token* in AWS Parameter Store and make it available to our Spring Boot application that is deployed to ECS Fargate.
+
+## Pre-reqs
+
+What you need to get started, or most likely, what you already have in place:
+
+- A **Spring Boot** application 
+- Deployed on **AWS**
+
+
+## Store your secrets in AWS Parameter Store
+
+Open up AWS Systems Manager then go to Parameter Store under Application Management.
+
+- Click `Create Parameter`.
+- In the *name* field, enter `/config/application/github.token`.  
+- In the *description* field, enter `GitHub API Token`
+- Select `SecureString` from the *type* field.  Accept the default KMS key source.
+- In the *text* field, enter `ABC123`.  
+- Click `Create Parameter` to save it.
+
+A few things to note, for the *type* field we could of used a String but instead I chose SecureString to keep the value encrypted at rest.  You can do either and it will be transparently decrypted when it is retrieved.  For the `name` field, the prefix `/config/application/` is important but everything else is just made up.  I'll go into why the prefix is important in just a bit.
+
+
+## Retrieving Secrets
+
+The "secret" sauce of this solution is in the integration.  Now that your configuration is securely stored in AWS, you need a good way to retrieve it.  A poor way to integrate would be to specifiy each property in the ECS task definition using ValueFrom mappings.  That's just more configuration and more to maintain.  Unsurprisingly, there's a great Spring integration that can help us out called **Spring Cloud Starter AWS Parameter Store Config**.  

```

Comment:
Now that your configuration is securely stored in AWS, you need a `reliable` way to retrieve it.